src/coreclr/nativeaot/System.Private.CoreLib/src/System/AppContext.NativeAot.cs

```diff
@@ -40,18 +42,60 @@ internal static void OnUnhandledException(object e)
 
         private static unsafe string GetRuntimeModulePath()
         {
-            // We aren't going to call this method, we just need an address that we know is in this module.
-            // As this code is NativeAOT only, we know that this method will be AOT compiled into the executable,
-            // so the entry point address will be in the module.
-            void* ip = (void*)(delegate*<string>)&GetRuntimeModulePath;
-            if (RuntimeAugments.TryGetFullPathToApplicationModule((nint)ip, out _) is string modulePath)
+            return RuntimeModulePathHolder.RuntimeModulePath;
+        }
+
+        internal static class RuntimeModulePathHolder
+        {
+            private static string? s_workingDirectoryAtStartup;
+            private static string WorkingDirectoryAtStartup
+            {
+                // In case a user's module initializer tries to get AppContext.BaseDirectory
+                // before our module initializer has run, we need to set up the getter to initialize
+                // the value.
+                get => s_workingDirectoryAtStartup ??= Environment.CurrentDirectory;
+            }
+
+            #pragma warning disable CA2255 // The 'ModuleInitializer' attribute is only intended to be used in application code or advanced source generator scenarios
+                                           // We need to capture this value at startup in case the user changes Environment.CurrentDirectory
+                                           // before the first execution of AppContext.BaseDirectory.
+            [ModuleInitializer]
+            #pragma warning restore CA2255
+            public static void Initialize()
             {
-                return modulePath;
+                s_workingDirectoryAtStartup = Environment.CurrentDirectory;
```

Comment:
Does this mean that close to every NAOT app is going to compute current directory during startup now?